cmd/testrunner/sensei.go

```diff
@@ -0,0 +1,243 @@
+			adminConn, err = grpc.Dial(rpcHost, opts...)
+			if err != nil {
+				return err
+			}
+			return err
+		}, 60)
+	if err != nil {
+		return nil, err
+	}
+	defer adminConn.Close()
+
+	logger.Infow("Creating node")
+
+	adminClient := sensei.NewAdminClient(adminConn)
+
+	createResp, err := adminClient.CreateAdmin(context.Background(), &sensei.CreateAdminRequest{
+		Username:   "admin",
+		Alias:      alias,
+		Passphrase: "secret",
+		Start:      true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	pubkey, err := route.NewVertexFromStr(createResp.Pubkey)
+	if err != nil {
+		return nil, err
+	}
+
+	time.Sleep(time.Second)
+
+	mac := createResp.Macaroon
+
+	cred, err := NewMacaroonCredential(mac)
+	if err != nil {
+		return nil, err
+	}
+
+	opts = append(opts,
+		grpc.WithPerRPCCredentials(cred),
+	)
+
+	logger.Infow("Connecting with macaroon")
+
+	var conn *grpc.ClientConn
+	err = tryFunc(
+		func() error {
+			var err error
+			conn, err = grpc.Dial(rpcHost, opts...)
+			if err != nil {
+				return err
+			}
+			return err
+		}, 60)
+	if err != nil {
+		return nil, err
+	}
+
+	return &senseiConnection{
+		alias:  alias,
+		conn:   conn,
+		pubKey: pubkey,
+		client: sensei.NewNodeClient(conn),
+	}, nil
+}
+
+func (l *senseiConnection) PubKey() string {
+	return l.pubKey.String()
+}
+
+func (l *senseiConnection) Close() {
+	l.conn.Close()
+}
+
+func (l *senseiConnection) Connect(key, host string) error {
+	_, err := l.client.ConnectPeer(context.Background(), &sensei.ConnectPeerRequest{
+		NodeConnectionString: fmt.Sprintf("%v@%v", key, host),
+	})
+
+	return err
+}
+
+func (l *senseiConnection) NewAddress() (string, error) {
+	resp, err := l.client.GetUnusedAddress(
+		context.Background(),
+		&sensei.GetUnusedAddressRequest{},
+	)
+	if err != nil {
+		return "", err
+	}
+	return resp.Address, nil
+}
+
+func (l *senseiConnection) OpenChannel(peerKey string, amtSat int64,
+	pushAmtSat int64, private bool) error {
+
+	_, err := l.client.OpenChannels(context.Background(), &sensei.OpenChannelsRequest{
+		Channels: []*sensei.OpenChannelInfo{
+			{
+				AmtSatoshis: uint64(amtSat),
+				Public:      !private,
+
+				// TODO:
+				// NodePubkeyString:   peerKey,
+				// SpendUnconfirmed:   true,
+				// PushSat:            pushAmtSat,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *senseiConnection) SetPolicy(chanPoint *lnrpc.ChannelPoint,
+	policy *graphreader.PolicyData) error {
+
+	// Not necessary for sender node.
+	panic("not implemented")
+}
+
+func (l *senseiConnection) ActiveChannels() (int, error) {
+	resp, err := l.client.ListChannels(context.Background(), &sensei.ListChannelsRequest{})
+	if err != nil {
+		return 0, err
+	}
+
+	// TODO: How to get numer of active channels?
+	return len(resp.Channels), nil
+}
+
+func (l *senseiConnection) IsSynced(totalEdges, localChannels int) (bool, error) {
+	// TODO: How to get number of edges received through gossip + local edges?
+
+	panic("not implemented")
```

Comment:
I just added a NetworkGraphInfo endpoint you can use to fetch number of channels + nodes in the graph.  I'm not actually sure if this includes local edges -- you might need to manually adjust by listing channels as well.

Reply:
Okay. Yes, so what the benchmark needs is some sort of 'completed' signal when all channel updates have been received. Not just having the channels isn't sufficient, because without the policies a route can not yet be calculated.

Would it be possible to add a counter for that? Something that counts that number of known edge policies.

Reply:
Ah ok.  ldk's graph keeps track of info for the channel policy in each direction.  do you want me to count known policies in each direction separately? so 1 channel with 2 known policies = 2 known edge policies?  or do you want me to count a "known edge" only if both policies for a channel are known?

Reply:
I added a num_known_edge_policies count to the NetworkGraphInfo response.  It counts a known policies in each direction independently.  So a single channel where both policies are known will return a count of 2.  let me know if you need something different.

Reply:
No, that's exactly what I need. Thank you